### ISSUE TITLE
Move "Account" model to own Django app

### ISSUE THREAD

**Issue by brylie** [CiviWiki/OpenCiviWiki#855]
Currently, all of our models are defined under the `api` app. We plan to re-organize the code structure to follow a more conventional Django project structure and deprecate the frontend SPA.

Move the account-related code/models into a new `accounts` app.

## Related code
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/civiwiki/urls.py#L24

https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/civiwiki/urls.py#L30

### Create custom user model
We can take this opportunity to create a custom user model. The project currently uses the default Django user model, so our custom user model can just inherit from `django.contrib.auth.models.User`

https://github.com/CiviWiki/OpenCiviWiki/blob/283460fef79dc69dc526c7ef5c29e4ec82bb9874/project/authentication/authentication.py#L3

See the Django docs for [customizing authentication with a substitute user model](https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#substituting-a-custom-user-model).'

Update the `Account` model to use the custom user model, and move the `Account` model to the `accounts` app:
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/api/models/account.py#L108

Make sure to update our registration form to use the correct user model:
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/authentication/forms.py#L40-L42

### Use Django class-based authentication views
For the most part, all authentication logic is already written for us. Remove the custom authentication views and replace them with the built-in, [class-based authentication views](https://docs.djangoproject.com/en/3.2/topics/auth/default/#module-django.contrib.auth.views).

### User related forms and templates
We need to make sure users can register and change their passwords, at the very least. So, we need to create templates to render the registration forms.

https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/authentication/forms.py#L20

Our current registration form code may be a bit bloated, such as the sections on password validation which can otherwise be provided by Django password validators.

Remove the following:
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/authentication/forms.py#L35-L38
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/authentication/forms.py#L77-L84

This may be redundant:
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/authentication/forms.py#L44-L50

### Passwords
Use Django built-in [password validators](https://docs.djangoproject.com/en/3.2/ref/settings/#std:setting-AUTH_PASSWORD_VALIDATORS) instead of custom code:
https://github.com/CiviWiki/OpenCiviWiki/blob/f152ad5986ebfd25ff60ad5ca7c34563818f1e17/project/authentication/forms.py#L66-L75

### CSRF
We probably need to re-enable cookie-based CSRF tokens, or use whatever is Django default:

https://github.com/CiviWiki/OpenCiviWiki/blob/283460fef79dc69dc526c7ef5c29e4ec82bb9874/project/civiwiki/settings.py#L60-L62

Consider whether it is desirable to disable CSRF on REST API views (or whether we will still support REST authentication after this pull request):
https://github.com/CiviWiki/OpenCiviWiki/blob/ff74546f8fdc277c9bc5705f3df0c807c1130c62/project/api/authentication.py#L5

**Comment by blackpandan** [comment]
hello I will be happy to contribute to this project, I might need your guidance

**Comment by brylie** [comment]
Thanks @blackpandan! Please let me know what questions or challenges you encounter when getting started.

**Comment by brylie** [comment]
@blackpandan I am available for the next hour in a [CiviWiki office hours hangout](https://codebuddies.org/hangout/qmdQirv9AhdNYkyGo) on Codebuddies.org.

**Comment by brylie** [comment]
@blackpandan please be aware of issue #856 as it has overlap with the accounts app/model.

**Comment by brylie** [comment]
@blackpandan I have included some further details in the issue description to get a better sense of the scope of this task. Please let me know your thoughts or concerns.

**Comment by brylie** [comment]
@blackpandan I'm just checking in. Is there any way I can be of assistance with this task?

Make sure to work with the latest development branch, since some of the work in this task has already been done in other tasks.

**Comment by a7madnassar** [comment]
I would like to help with this task if there's any work left to be done.

**Comment by brylie** [comment]
Thanks, @a7madnassar I'll assign you to the issue. Please keep me updated with any progress or difficulties. I'm available to help.

**Comment by a7madnassar** [comment]
Great, I was just wondering what is left in this issue. Also, some mentioned files do not exist in the "develop" branch. For example the settings.py file. Can you point me in the right direction here? Thanks.

**Comment by brylie** [comment]
The `settings.py` file is in the `/project/civiwiki/` directory.

I updated the description to remove some parts that have already been done. The remaining tasks should refer to the `accounts` app, which is located in `/project/accounts/`.

E.g. create the custom user model in the Accounts app.

**Comment by a7madnassar** [comment]
I forked the "develop" branch to work on. However, this branch does not include the /project/civiwiki directory in it. Which branch should I fork for the modifications?

**Comment by a7madnassar** [comment]
I guess it should be the "all-contributors/add-fardeenes7" branch, right? Also, for the custom user, are we extending django.contrib.auth.models.User or django.contrib.auth.models.AbstractUser?

**Comment by brylie** [comment]
My bad. The `develop` branch has `/project/core/` directory where you can find settings. The `/project/accounts/` directory is the destination for all accounts-related code and templates.

**Comment by a7madnassar** [comment]
Great, that makes it clear now. Thanks.

**Comment by a7madnassar** [comment]
I started working on this issue and was able to create a custom User class that extends django.contrib.auth.models.User. This means it will be just a proxy model. I also investigated the rest of the tasks and most of them should be OK except for the main task of moving the Account model to the new accounts app. I will keep this task as the last one because currently if I try to drop and add the new model with the same name, it breaks migrations. Do you have any recommendations on how to approach moving the model to the accounts app?

**Comment by brylie** [comment]
@a7madnassar please make a pull request with your current changes. That way, we can review your code already at this point and possibly help resolve/avoid some issues. Once you have opened a pull request, I will try and help with the migrations issue.

**Comment by a7madnassar** [comment]
@brylie I just submitted a pull request for the work I have done so far. Here's a summary of what I did:
1. Swapped the default User model for a custom model that extends django.contrib.auth.models.AbstractUser.
2. Implemented registration and login logic using class-based views under a new '/accounts/' path. Users currently can register and log in.

I chose to postpone creating the new Account model to the end as it is a little more involved.
